File: week_1/code/subtree_of_another_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function (root, subRoot) {
    if (!subRoot) return true;
    if (!root) return false;

    if (isSameTree(root, subRoot)) return true;

    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};

var isSameTree = function (p, q) {
    if (!p && !q) return true;
    if (!p || !q || p.val !== q.val) return false;
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};

// Helper to build tree from array
function buildTree(arr) {
    if (!arr.length) return null;
    let root = new TreeNode(arr[0]);
    let queue = [root];
    let i = 1;
    while (queue.length > 0 && i < arr.length) {
        let node = queue.shift();
        if (i < arr.length && arr[i] !== null) {
            node.left = new TreeNode(arr[i]);
            queue.push(node.left);
        }
        i++;
        if (i < arr.length && arr[i] !== null) {
            node.right = new TreeNode(arr[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

// Test cases
// Test 1: root = [3,4,5,1,2], subRoot = [4,1,2] (True)
let root1 = buildTree([3, 4, 5, 1, 2]);
let subRoot1 = buildTree([4, 1, 2]);
console.log("Test 1:", isSubtree(root1, subRoot1), "(Expected: true)");

// Test 2: root = [3,4,5,1,2,null,null,null,null,0], subRoot = [4,1,2] (False)
let root2 = buildTree([3, 4, 5, 1, 2, null, null, null, null, 0]);
let subRoot2 = buildTree([4, 1, 2]);
console.log("Test 2:", isSubtree(root2, subRoot2), "(Expected: false)");

File: week_1/code/subtree_of_another_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        """
        Given the roots of two binary trees root and subRoot, return true if there is a subtree of root with the same structure and node values of subRoot and false otherwise.
        """
        if not subRoot: return True
        if not root: return False
        
        if self.isSameTree(root, subRoot):
            return True
            
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
    
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q: return True
        if not p or not q or p.val != q.val: return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

# Helper to build tree from level order list
def buildTree(arr):
    if not arr: return None
    root = TreeNode(arr[0])
    queue = [root]
    i = 1
    while queue and i < len(arr):
        node = queue.pop(0)
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root

# Test cases
if __name__ == "__main__":
    sol = Solution()
    
    # Test 1: root = [3,4,5,1,2], subRoot = [4,1,2] (True)
    root1 = buildTree([3,4,5,1,2])
    subRoot1 = buildTree([4,1,2])
    print(f"Test 1: {sol.isSubtree(root1, subRoot1)} (Expected: True)")

    # Test 2: root = [3,4,5,1,2,null,null,null,null,0], subRoot = [4,1,2] (False)
    root2 = buildTree([3,4,5,1,2,None,None,None,None,0])
    subRoot2 = buildTree([4,1,2])
    print(f"Test 2: {sol.isSubtree(root2, subRoot2)} (Expected: False)")
